c10/metal/special_math.h

```diff
@@ -503,5 +503,27 @@ inline T spherical_bessel_j0(T x) {
   return static_cast<T>(::metal::sin(x) / x);
 }
 
+// TODO (dcci): There are currently two implementations on log1p,
+// one here (written using a Metal shader), the other in BinaryKernels.mm,
+// that uses MPS. We should consider consolidating, maybe rewriting the
+// MPS implementation using this approximation, or the other way around.
+template <typename T>
+inline T log1p(T x) {
```

Comment:
Does metal really not have a more numerically stable way of calculating this built in? While mathematically it's equivalent to shifting it by one, my understanding is that the actual function implementation is way more stable when shifted in CPU/CUDA. This is a decent shim, but may cause numerical stability in actual code use cases.

Reply:
Do you have any suggestions on how to implement it in a more stable way?

The current MPS implementation doesn't seem to be that different

```

MPSGraphTensor* log1p(MPSGraph* mpsGraph, MPSGraphTensor* inputTensor) {
  MPSGraphTensor* oneTensor = [mpsGraph constantWithScalar:1.0 dataType:inputTensor.dataType];
  MPSGraphTensor* addedTensor = [mpsGraph additionWithPrimaryTensor:inputTensor secondaryTensor:oneTensor name:nil];
  return [mpsGraph logarithmWithTensor:addedTensor name:nil];
}
```

[literally just adding 1 to the tensor and computing the log].

Maybe we can leave this as a follow-up?